### ISSUE TITLE
APP CRASH signal arriving during dr_app_setup races with handler removal

### ISSUE THREAD

**Issue by derekbruening** [DynamoRIO/dynamorio#2335]
Today DR installs its signal handlers during dr_app_setup(), but then we
remove them at the end of setup due to the split setup-vs-start design
(which like the stop-vs-detach design is inherently problematic...).  We
re-instate them during dr_app_start().  We thus have a race where a signal
such as SIGPROF can arrive while our handler is in place during setup, be
delivered to a native thread who has no TLS, and before the SIGSEGV from
the TLS safe read is raised, setup finishes and
os_process_not_under_dynamorio() reinstates the app handler.  We end up
with an app crash.

To solve: for dr_app_setup(), do not put our handlers in place, right?
signal_reinstate_handlers() will re-acquire the app handlers.
Can we also remove the not_under call at the end of setup?
Right now we have init doing "under", end of init doing "not under", and
start doing "under".  Maybe we can remove the first two.

Should we only support dr_app_setup_and_start() instead of the split?

**Comment by derekbruening** [comment]
Problem: we remove the init-time crash handlers when we set up the thread,
so we miss later init-time crashes like -unsafe_crash_process and
client-init crashes.  We could move the init-time remove, for dr_api_entry,
so some later point: a new routine we call at the end of dr_app_setup() or
something.  We'll also have to check the init-time-stored app handler for
-invoke_app_on_crash.

We also need to not swap the signal mask either: except for the init-time
crash handlers.  In fact: we're not swapping the mask on start/stop
transitions at all today!  Should probably be filed as a separate issue.

**Comment by derekbruening** [comment]
Re-opening due to problems with client threads and init-time signals:

In 053ee07 I added a call to signal_reinstate_handlers() in a client
thread's run function -- but it shouldn't be needed there as the client
thread shares the app thread group's signal handler table, and having it
there breaks i#2335 if a client thread is created during init.

Apparently I added it to get the client.signal test to pass without
thinking it through.  That test creates a client thread and sends it a
signal during dr_client_main().  But w/ this #2335 change DR's signal
handlers (except init-time crash handlers) are not set up then.

The erroneous call also prevents client.thread from failing: that test
makes a client thread during init who sets up an itimer, but with no
handler, the client thread just dies, causing the parent to wait forever.

This is all just due to the messiness of init vs start.  It seemed clean to
move everything to start, but that breaks clients making threads during
init -- and init time is the natural time for a client thread to be
created and for its own itimer to be started.

Possible solutions:

A) Stop supporting os_process_not_under_dynamorio().

   This means we'd stop supporting dr_app_setup() and require
   dr_app_setup_and_start() and we'd similarly drop support for
   dr_app_stop(), and move handler init earlier again.  We would also have
   to not support -go_native_at_bb_count or any other use of
   os_process_not_under_dynamorio().

   We could potentially qualify the lack of support to only be when there
   are client threads.

B) Delay various aspects of client threads until start, and suspend/block
   those aspects during os_process_not_under_dynamorio().  We'd document
   this, including that signals can't be received by client threads until
   the app starts.

   For delaying, some options:

   1) Do not support client thread creation during init.  This seems really
   annoying to client writers.  For os_process_not_under_dynamorio() we'd
   have to suspend client threads by sending a signal.

   2) Delay client thread itimers set up during init until start, and disable
   client thread itimers in os_process_not_under_dynamorio() -- except that
   is difficult to do as it has to be done from the client thread context:
   we'd have to send a special signal and wait for completion before we
   removed our handlers.

   3) Delay client thread execution until start via a wait in the main run
   function.  We'd also want to suspend client threads for
   os_process_not_under_dynamorio() by sending a signal.

Also, what if a client wants to send itself a signal during init (xref DR
doing this to get xstate info)?  We should document what's supported there.

**Comment by derekbruening** [comment]
Given that there are use cases that want setup to remain split from start (including code cache prepopulation), the plan is to implement B.3.  The part of suspending client threads is split of as #2784.

**Comment by Carrotman42** [comment]
PR #2785 does not correctly initialize `dr_app_started` before the `dr_client_main` of a client is called, so it contains a race where a client thread tries to wait on a NULL `dr_app_started` event and causes a crash. A simple fix is to just move the initialization of this variable to before clients are initialized during `instrument_init`.
